### PR TITLE
adding ecr policy resource

### LINKED ISSUE #535: Assigning ECR repository permission
### What problem are you facing?
The current [ecr repository object](https://doc.crds.dev/github.com/crossplane/provider-aws/ecr.aws.crossplane.io/Repository/v1alpha1@v0.16.0) is not able to assign permission on the ecr repo. In cross account environments this is needed because setting permissions on the repository is the easiest way to share repositories. This enables especially cross account CICD setups.

### How could Crossplane help solve your problem?
Setting these permissions could be implemented either within the existing repository object or (similar to [bucketPolicy](https://doc.crds.dev/github.com/crossplane/provider-aws/s3.aws.crossplane.io/BucketPolicy/v1alpha2@v0.16.0) for S3) as addtional object.

See here the terraform [ecr_repository_policy](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/ecr_repository_policy) resource for example.
The [aws sdk](https://github.com/aws/aws-sdk-go/blob/v1.37.7/models/apis/ecr/2015-09-21/docs-2.json#L31) supports that as well.

### PR DESCRIPTION
**Pull request by enderv**
Signed-off-by: chris <cvodak@ea.com>

### Description of your changes

Fixes #535
Adds support for ECR Policies by copying the same approach as S3 Bucket policy where you can either do raw string or a struct based policy. 

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested
Run automated tests. Tested by adding example policy to a repo created using a repo resource. Modified the policy to make sure changes went through.

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

File: examples/ecr/respositorypolicy-json.yaml
Comment (chlunde) [inline examples/ecr/respositorypolicy-json.yaml:11]:
```suggestion
    rawPolicy:  |
```

Just skimmed the PR and noticed this, I have not reviewed it :)

Comment (enderv) [reply examples/ecr/respositorypolicy-json.yaml]: doh, nice catch.

File: pkg/controller/ecr/repositorypolicy/controller.go
Comment (hasheddan) [inline pkg/controller/ecr/repositorypolicy/controller.go:114]: I don't think we would ever hit `ecr.IsPolicyNotFoundErr` here right?

**Review by hasheddan — CHANGES_REQUESTED** [review]

File: pkg/controller/ecr/repositorypolicy/controller.go
Comment (enderv) [reply pkg/controller/ecr/repositorypolicy/controller.go:114]: yep. will fix

File: pkg/controller/ecr/repositorypolicy/controller.go
Comment (enderv) [reply pkg/controller/ecr/repositorypolicy/controller.go:118]: thats my understanding.

File: pkg/controller/ecr/repositorypolicy/controller.go
Comment (enderv) [reply pkg/controller/ecr/repositorypolicy/controller.go:116]: Oh good point. yep will change.

File: pkg/controller/ecr/repositorypolicy/controller.go
Comment (enderv) [reply pkg/controller/ecr/repositorypolicy/controller.go:124]: Doh, had to double check, thought there wasn't any but looks like two fields will get em in.

**Comment by enderv** [comment]
@hasheddan think I got everything, not sure how you normally do this just resolved as I went through.

**Comment by HerrmannHinz** [comment]
niceone @enderv

File: pkg/controller/ecr/repositorypolicy/controller.go
Comment (hasheddan) [reply pkg/controller/ecr/repositorypolicy/controller.go]: @enderv I think are some places where we do this with the assumption that it will never be `nil`. I personally prefer to check that it is not `nil` as, though it wouldn't be great by SDK maintainers to change this behavior, it could be changed and we would then have a potential panic here. What do you think?

Comment (enderv) [reply pkg/controller/ecr/repositorypolicy/controller.go]: sounds good to me. On it

Comment (enderv) [reply pkg/controller/ecr/repositorypolicy/controller.go]: Oh actually already got this covered with my lateinitialization change have a nil check in the function there and now just comparing the `RepositoryPolicyParameters`

Comment (hasheddan) [reply pkg/controller/ecr/repositorypolicy/controller.go]: Aren't we still deferencing before late init?

https://github.com/crossplane/provider-aws/pull/573/files#diff-b54d7ac2e4ab0e950fd8fb8c791b11196da72121140174398d12c779bc643624R116

Comment (enderv) [reply pkg/controller/ecr/repositorypolicy/controller.go]: :eyes: so we are. on it.

File: apis/ecr/v1alpha1/policy_types.go
Comment (muvaf) [inline apis/ecr/v1alpha1/policy_types.go:2]:
```suggestion
Copyright 2021 The Crossplane Authors.
```

in all files.

File: apis/ecr/v1alpha1/referencers.go
Comment (muvaf) [inline apis/ecr/v1alpha1/referencers.go:46]:
```suggestion
		return errors.Wrap(err, "spec.forProvider.repositoryName")
```

File: apis/ecr/v1alpha1/referencers.go
Comment (muvaf) [inline apis/ecr/v1alpha1/referencers.go:52]:
```suggestion
	if mg.Spec.ForProvider.Policy != nil {
```

`for` would be no-op if array is `nil` or has no elements.

File: apis/ecr/v1alpha1/referencers.go
Comment (muvaf) [inline apis/ecr/v1alpha1/referencers.go:75]:
```suggestion
```
This `nil` check is taken care of by the resolver.

File: apis/ecr/v1alpha1/referencers.go
Comment (muvaf) [inline apis/ecr/v1alpha1/referencers.go:84]:
```suggestion
				return errors.Wrap(err, fmt.Sprintf("spec.forProvider.statements[%d].principal.awsPrincipals[%d].iamUserArn", statementIndex, i))
```

File: apis/ecr/v1alpha1/referencers.go
Comment (muvaf) [inline apis/ecr/v1alpha1/referencers.go:99]:
```suggestion
				return errors.Wrap(err, fmt.Sprintf("spec.forProvider.statements[%d].principal.awsPrincipals[%d].iamRoleArn", statementIndex, i))
```

File: examples/ecr/repositorypolicy.yaml
Comment (muvaf) [inline examples/ecr/repositorypolicy.yaml:10]:
```suggestion
    repositoryNameRef:
      name: example
```

File: pkg/controller/ecr/repositorypolicy/controller.go
Comment (muvaf) [inline pkg/controller/ecr/repositorypolicy/controller.go:107]:
```suggestion
```

I think that's a valid error case - users cannot set policy of a repository that doesn't exist. We ignore `PolicyNotFoundErr` because this controller will create the policy.

File: pkg/controller/ecr/repositorypolicy/controller.go
Comment (muvaf) [inline pkg/controller/ecr/repositorypolicy/controller.go:127]:
```suggestion
```

You can return `ResourceLateInitialized: !cmp.Equal(current, &cr.Spec.ForProvider)` now with the latest crossplane-runtime.

File: pkg/controller/ecr/repositorypolicy/controller.go
Comment (muvaf) [inline pkg/controller/ecr/repositorypolicy/controller.go:143]:
```suggestion
```

The generic reconciler will set this now automatically.

File: pkg/controller/ecr/repositorypolicy/controller.go
Comment (muvaf) [inline pkg/controller/ecr/repositorypolicy/controller.go:146]:
```suggestion
```

File: pkg/controller/ecr/repositorypolicy/controller.go
Comment (muvaf) [inline pkg/controller/ecr/repositorypolicy/controller.go:181]:
```suggestion
```

The generic reconciler will set this now automatically.

File: pkg/controller/ecr/repositorypolicy/controller.go
Comment (muvaf) [inline pkg/controller/ecr/repositorypolicy/controller.go:193]:
```suggestion
func (e *external) rawPolicyData(original *v1alpha1.RepositoryPolicy) (string, error) {
```

nitpick: returning `string` could make it easier to work with. We'd use `awsclients.String(ref)` as usual whenever we send it to AWS. And `formatRepositoryPolicy` feels like there is an in-place change but what we do is mostly conversion from one format to another.

Comment (muvaf) [reply pkg/controller/ecr/repositorypolicy/controller.go:193]: Also what do you think about moving this to `pkg/clients/ecr/repository_policy.go`?

File: pkg/controller/ecr/repositorypolicy/controller.go
Comment (muvaf) [inline pkg/controller/ecr/repositorypolicy/controller.go:188]:
```suggestion
```

File: pkg/controller/ecr/repositorypolicy/controller.go
Comment (muvaf) [inline pkg/controller/ecr/repositorypolicy/controller.go:172]:
```suggestion
	return managed.ExternalUpdate{}, awsclient.Wrap(err, errUpdate)
```

File: pkg/controller/ecr/repositorypolicy/controller.go
Comment (muvaf) [inline pkg/controller/ecr/repositorypolicy/controller.go:155]:
```suggestion
	return managed.ExternalCreation{}, awsclient.Wrap(err, errCreate)
```

Is there a specific reason that requires us to issue a spec update?

File: pkg/controller/ecr/repositorypolicy/controller.go
Comment (enderv) [reply pkg/controller/ecr/repositorypolicy/controller.go:193]: kk and i'll go ahead and move it as well

File: pkg/controller/ecr/repositorypolicy/controller.go
Comment (enderv) [reply pkg/controller/ecr/repositorypolicy/controller.go]: Thought there was but looking again can't seem to figure out why will remove that.

File: apis/ecr/v1alpha1/doc.go
Comment (muvaf) [inline apis/ecr/v1alpha1/doc.go:2]: We use the current year in only the files created with that PR.

File: apis/ecr/v1alpha1/referencers.go
Comment (muvaf) [inline apis/ecr/v1alpha1/referencers.go:89]:
```suggestion
```

File: apis/ecr/v1alpha1/register.go
Comment (muvaf) [inline apis/ecr/v1alpha1/register.go:2]:
```suggestion
Copyright 2020 The Crossplane Authors.
```

File: pkg/clients/ecr/repository_policy.go
Comment (muvaf) [inline pkg/clients/ecr/repository_policy.go:48]:
```suggestion
	in.RegistryID = awsclient.LateInitializeStringPtr(in.RegistryID, r.RegistryId)
```

We can't override spec if it already exists.

File: apis/ecr/v1alpha1/policy_types.go
Comment (muvaf) [inline apis/ecr/v1alpha1/policy_types.go:28]:
```suggestion
	// Region is the region you'd like your RepositoryPolicy to be created in.
```

**Review by muvaf — APPROVED** [review]
Looking pretty good, should be ready to merge after a few quick fixes and commit squashing. `IsUpToDate` logic seems a bit risky, have you tested to see whether it returns false negatives?

File: apis/ecr/v1alpha1/doc.go
Comment (enderv) [reply apis/ecr/v1alpha1/doc.go]: oops

**Comment by enderv** [comment]
@muvaf got the rest of those changes in and squashed. For the `IsUpToDate` did a few checks just changing ordering of the json to make sure equivalent came through correct. Added another test but if anything else in mind to check just let me know.

**Review by hasheddan — APPROVED** [review]

**Comment by muvaf** [comment]
Thank you @enderv !!

**Pull request merged**
